### PR TITLE
Fix clang-tidy C++17 checks

### PR DESCRIPTION
**Pull request by Noordfrees**
- `modernize-use-nodiscard`: Functions that only perform a calculation and return the result without any side-effects should be annotated `[[nodiscard]]`. This annotation causes compilers to emit a warning when you invoke the function without using its result.
- `readability-redundant-declaration`: Sort of what the includes script already checks, but more sophisticated so it catches more occurrences. (The includes script also does something else which clang-tidy can't.)
- `modernize-concat-nested-namespaces`: More concise and readable declarations of nested namespaces.
- Re #5641

### CODE REVIEW

**Comment by tothxa** [comment]
> * `modernize-use-nodiscard`: Functions that only perform a calculation and return the result without any side-effects should be annotated `[[nodiscard]]`. This annotation causes compilers to emit a warning when you invoke the function without using its result.

This is quite ugly. And if clang-tidy can find them, compilers should also be able to do it without annotations.

**Comment by hessenfarmer** [comment]
> > * `modernize-use-nodiscard`: Functions that only perform a calculation and return the result without any side-effects should be annotated `[[nodiscard]]`. This annotation causes compilers to emit a warning when you invoke the function without using its result.
> 
> This is quite ugly. And if clang-tidy can find them, compilers should also be able to do it without annotations.

I agree it is ugly on the first view. However I always needed some time to accomodate with the new syntaxes enforced by clang_tidy to see their benefit. In this case this is something like a special comment for me that reminds new coders or not people with limited knowledge like myself to always use the result of such a function call, which is not that bad imho.

File: src/ai/defaultai.cc [hunk 1]
```diff
@@ -83,15 +83,6 @@ constexpr int kMineTypes = 4;
 namespace AI {
 
 // Fix undefined references
```
Comment (hessenfarmer) [inline src/ai/defaultai.cc:85]: I am not sure whether the solution for these values would be so easy. I think solving this issue would need more thoughts as the comment implies that without these declarations the compiler would complaion about missing references. In the header file these values are declared as private members of the defaultAI struct, but maybe they are not always used only in a private context.

**Comment by tothxa** [comment]
> In this case this is something like a special comment for me that reminds new coders or not people with limited knowledge like myself to always use the result of such a function call, which is not that bad imho.

The problem is that we don't check header files with clang-tidy, so these annotations only go to the `.cc` files, while they would be much better in the headers for this use case.

**Comment by Noordfrees** [comment]
Turns out we can fix headers too ;) Something that should be done for all other checks as well in a follow-up branch.

File: src/ai/defaultai.cc (hunk 1 above)
Comment (Noordfrees) [reply src/ai/defaultai.cc:85]: The compiler warning here says
```
warning: out-of-line definition of constexpr static data member is redundant in C++17 and is deprecated [-Wdeprecated]
```
C++14 and older needs this, C++17 is smarter and makes this superfluous. And apparently it compiles and links fine without them.

**Comment by tothxa** [comment]
> Turns out we can fix headers too ;) Something that should be done for all other checks as well in a follow-up branch.

[I'd told you before](https://github.com/widelands/widelands/discussions/5262#discussioncomment-2212007). :)

**Comment by Noordfrees** [comment]
> > Turns out we can fix headers too ;) Something that should be done for all other checks as well in a follow-up branch.
> 
> [I'd told you before](https://github.com/widelands/widelands/discussions/5262#discussioncomment-2212007). :)

I must have missed that :/ Will implement these suggestions in the next branch

File: src/ai/defaultai.cc (hunk 1 above)
Comment (hessenfarmer) [reply src/ai/defaultai.cc:85]: ok. did not know that. I just ran into issues with these definitions a lot of times, so I was alerted. So I am glad that c++17 is better in this.

**Review by hessenfarmer — APPROVED** [review]
well, I checked the changes and they seem to be good. and it builds fine.    
So ok from my side.

**Pull request merged**